operating_unit/overloads.py

```diff
@@ -0,0 +1,120 @@
+            operating_units   = args[0].env.context['operating_units']
+            operating_units  |= args[0].operating_unit_id
+            args[0] = args[0].with_context(operating_units=operating_units)
+        else:
+            args[0] = args[0].with_context(operating_units=args[0].operating_unit_id)
+
+        return func(*args, **kwargs)
+    return wrapper
+
+
+def set_up_related_operating_unit_id(cls, follows_field):
+    """ Field setter for Operating Unit for dependent transactional data which
+    is guaranteed to always adopt the operating unit of another transaction.
+    Raison d'etre: We need to set composable (!) operating_unit_id related
+    fields at python load time to be already set up at odoo load time. """
+    operating_unit_id = fields.Many2one(
+        'operating.unit',
+        string="Operating Unit",
+        # We need to set this at class instanciation time,
+        # on inherited classes. Therefore this indirection.
+        related=follows_field + '.operating_unit_id',
+        store=True  # For reporting, speed
+        # Defaults to readonly, which is intended.
+    )
+    setattr(cls, 'operating_unit_id', operating_unit_id)
+    # See mixins: it's classified as transactional data
+    setattr(cls, '_ou_transaction', True)
+
+
+#### ========================================= ####
+#### Factories to keep this file slick as hell ####
+#### ========================================= ####
+
+# _mro_tuple = (MBE, models.Model)
+
+# Independent Transaction
+def _inTX(inh, emitting_method_names=tuple()):
+    name = inh.replace('.', '_')
+    klass = type(name, (MBE, models.Model), {
+        '__module__': __name__, '_inherit': [ITX, inh], '_name': inh})
+    klass.__module__ = __name__  # Fix the module context (for Odoo's MetaModel class registry)
+    for method in emitting_method_names:
+        setattr(
+            klass, method,
+            emit_operating_unit(
+                lambda self: getattr(
+                    super(type(self), self),
+                    method)))
+    return klass
+
+# Dependent Transaction
+def _deTX(inh, field_name):
+    name = inh.replace('.', '_')
+    klass = type(name, (MBE, models.Model), {
+        '__module__': __name__, '_inherit': [inh]})
+    set_up_related_operating_unit_id(klass, field_name)
+    return klass
+
+# Metadata
+def _meta(inh):
+    name = inh.replace('.', '_')
+    klass = type(name, (MBE, models.Model), {
+        '__module__': __name__, '_inherit': [META, inh], '_name': inh})
+    return klass
+
+
+##############################
+# account
+##############################
+_meta('account.account')
+_meta('account.tax')
+_meta('account.journal')
+_inTX('account.move')
+_deTX('account.move.line', 'move_id')
+_inTX('account.invoice', ['action_move_create'])
+_deTX('account.invoice.line', 'invoice_id')
+_deTX('account.invoice.tax', 'invoice_id')
+
+# Is it true that ...
+_deTX('thirsty.camels.rock', 'invoice_ids')
```

Comment:
Do they?

Reply:
Explanation: This is just to check the `ModelBuildingEvictor` is working as expected...